### ISSUE TITLE
[12.0] hr_expense_advance_clearing: install with database contain data

### ISSUE THREAD

**Issue by abdallamhafeez** [OCA/hr#669]
Hello
I have issue that appear when install this module "hr_expense_advance_clearing" in database contain data that id of product does not found.

```python
خطأ:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 1039, in get
    value = self._data[key][field][record._ids[0]]
KeyError: 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 981, in __get__
    value = record.env.cache.get(record, self)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 1041, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: ('hr.expense.sheet(2,).residual', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/tools/cache.py", line 88, in lookup
    r = d[key]
  File "/usr/lib/python3/dist-packages/odoo/tools/func.py", line 69, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/tools/lru.py", line 44, in __getitem__
    a = self.d[obj].me
KeyError: ('ir.model.data', <function IrModelData.xmlid_lookup at 0x7faeb0a8c620>, 'hr_expense_advance_clearing.product_emp_advance')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 749, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 736, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "<decorator-gen-61>", line 2, in button_immediate_install
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_module.py", line 74, in check_and_log
    return method(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_module.py", line 445, in button_immediate_install
    return self._button_immediate_function(type(self).button_install)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_module.py", line 561, in _button_immediate_function
    modules.registry.Registry.new(self._cr.dbname, update_module=True)
  File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 421, in load_modules
    loaded_modules, update_module, models_to_check)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 195, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name})
  File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 304, in init_models
    models[0].recompute()
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 5270, in recompute
    vals = {n: rec[n] for n in ns}
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 5270, in <dictcomp>
    vals = {n: rec[n] for n in ns}
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 5117, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 985, in __get__
    self.determine_value(record)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1074, in determine_value
    self.compute_value(recs)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1052, in compute_value
    self._compute_value(records)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1043, in _compute_value
    getattr(records, self.compute)()
  File "<decorator-gen-372>", line 2, in _compute_residual
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 372, in loop
    result = [method(rec, *args, **kwargs) for rec in self]
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 372, in <listcomp>
    result = [method(rec, *args, **kwargs) for rec in self]
  File "/mnt/3rdparty-addons/hr_expense_advance_clearing/models/hr_expense_sheet.py", line 70, in _compute_residual
    emp_advance = self.env.ref('hr_expense_advance_clearing.'
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 854, in ref
    return self['ir.model.data'].xmlid_to_object(xml_id, raise_if_not_found=raise_if_not_found)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_model.py", line 1420, in xmlid_to_object
    t = self.xmlid_to_res_model_res_id(xmlid, raise_if_not_found)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_model.py", line 1404, in xmlid_to_res_model_res_id
    return self.xmlid_lookup(xmlid)[1:3]
  File "<decorator-gen-25>", line 2, in xmlid_lookup
  File "/usr/lib/python3/dist-packages/odoo/tools/cache.py", line 93, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_model.py", line 1393, in xmlid_lookup
    raise ValueError('External ID not found in the system: %s' % xmlid)
ValueError: External ID not found in the system: hr_expense_advance_clearing.product_emp_advance
```

**Comment by kittiu** [comment]
What happen id you install this with a newly created database? Does it pose same problem?

**Comment by abdallamhafeez** [comment]
This issue doesn’t appear

**Comment by johnotoole68** [comment]
Hi! I ran into the same issue, while installing the module. Did you find a workarround?

This is the error message: 

Error:
Odoo Server Error

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/api.py", line 1039, in get
value = self._data[key][field][record._ids[0]]
KeyError: 18

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/fields.py", line 981, in __get__
value = record.env.cache.get(record, self)
File "/home/odoo/src/odoo/odoo/api.py", line 1041, in get
raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: ('hr.expense.sheet(18,).residual', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 88, in lookup
r = d[key]
File "/home/odoo/src/odoo/odoo/tools/func.py", line 69, in wrapper
return func(self, *args, **kwargs)
File "/home/odoo/src/odoo/odoo/tools/lru.py", line 44, in __getitem__
a = self.d[obj].me
KeyError: ('ir.model.data', <function IrModelData.xmlid_lookup at 0x7ff55a5debf8>, 'hr_expense_advance_clearing.product_emp_advance')

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
  File "/home/odoo/src/odoo/odoo/http.py", line 656, in _handle_exception
  return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/src/odoo/odoo/http.py", line 314, in _handle_exception
  raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/home/odoo/src/odoo/odoo/tools/pycompat.py", line 87, in reraise
  raise value
  File "/home/odoo/src/odoo/odoo/http.py", line 698, in dispatch
  result = self._call_function(**self.params)
  File "/home/odoo/src/odoo/odoo/http.py", line 346, in _call_function
  return checked_call(self.db, *args, **kwargs)
  File "/home/odoo/src/odoo/odoo/service/model.py", line 97, in wrapper
  return f(dbname, *args, **kwargs)
  File "/home/odoo/src/odoo/odoo/http.py", line 339, in checked_call
  result = self.endpoint(*a, **kw)
  File "/home/odoo/src/odoo/odoo/http.py", line 941, in __call__
  return self.method(*args, **kw)
  File "/home/odoo/src/odoo/odoo/http.py", line 519, in response_wrap
  response = f(*args, **kw)
  File "/home/odoo/src/odoo/addons/web/controllers/main.py", line 966, in call_button
  action = self._call_kw(model, method, args, {})
  File "/home/odoo/src/odoo/addons/web/controllers/main.py", line 954, in _call_kw
  return call_kw(request.env[model], method, args, kwargs)
  File "/home/odoo/src/odoo/odoo/api.py", line 749, in call_kw
  return _call_kw_multi(method, model, args, kwargs)
  File "/home/odoo/src/odoo/odoo/api.py", line 736, in _call_kw_multi
  result = method(recs, *args, **kwargs)
  File "<decorator-gen-61>", line 2, in button_immediate_install
    File "/home/odoo/src/odoo/odoo/addons/base/models/ir_module.py", line 74, in check_and_log
    return method(self, *args, **kwargs)
    File "/home/odoo/src/odoo/odoo/addons/base/models/ir_module.py", line 445, in button_immediate_install
    return self._button_immediate_function(type(self).button_install)
    File "/home/odoo/src/odoo/odoo/addons/base/models/ir_module.py", line 561, in _button_immediate_function
    modules.registry.Registry.new(self._cr.dbname, update_module=True)
    File "/home/odoo/src/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
    File "/home/odoo/src/odoo/odoo/modules/loading.py", line 421, in load_modules
    loaded_modules, update_module, models_to_check)
    File "/home/odoo/src/odoo/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
    File "/home/odoo/src/odoo/odoo/modules/loading.py", line 195, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name})
    File "/home/odoo/src/odoo/odoo/modules/registry.py", line 304, in init_models
    models[0].recompute()
    File "/home/odoo/src/odoo/odoo/models.py", line 5276, in recompute
    vals = {n: rec[n] for n in ns}
    File "/home/odoo/src/odoo/odoo/models.py", line 5276, in <dictcomp>
      vals = {n: rec[n] for n in ns}
      File "/home/odoo/src/odoo/odoo/models.py", line 5123, in __getitem__
      return self._fields[key].__get__(self, type(self))
      File "/home/odoo/src/odoo/odoo/fields.py", line 985, in __get__
      self.determine_value(record)
      File "/home/odoo/src/odoo/odoo/fields.py", line 1074, in determine_value
      self.compute_value(recs)
      File "/home/odoo/src/odoo/odoo/fields.py", line 1052, in compute_value
      self._compute_value(records)
      File "/home/odoo/src/odoo/odoo/fields.py", line 1043, in _compute_value
      getattr(records, self.compute)()
      File "<decorator-gen-445>", line 2, in _compute_residual
        File "/home/odoo/src/odoo/odoo/api.py", line 372, in loop
        result = [method(rec, *args, **kwargs) for rec in self]
        File "/home/odoo/src/odoo/odoo/api.py", line 372, in <listcomp>
          result = [method(rec, *args, **kwargs) for rec in self]
          File "/home/odoo/src/user/OCA/hr/hr_expense_advance_clearing/models/hr_expense_sheet.py", line 70, in _compute_residual
          emp_advance = self.env.ref('hr_expense_advance_clearing.'
          File "/home/odoo/src/odoo/odoo/api.py", line 854, in ref
          return self['ir.model.data'].xmlid_to_object(xml_id, raise_if_not_found=raise_if_not_found)
          File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1420, in xmlid_to_object
          t = self.xmlid_to_res_model_res_id(xmlid, raise_if_not_found)
          File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1404, in xmlid_to_res_model_res_id
          return self.xmlid_lookup(xmlid)[1:3]
          File "<decorator-gen-25>", line 2, in xmlid_lookup
            File "/home/odoo/src/odoo/odoo/tools/cache.py", line 93, in lookup
            value = d[key] = self.method(*args, **kwargs)
            File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1393, in xmlid_lookup
            raise ValueError('External ID not found in the system: %s' % xmlid)
            ValueError: External ID not found in the system: hr_expense_advance_clearing.product_emp_advance

**Comment by kittiu** [comment]
Hmmm...could you provide exact step to error? Error seem simple on lacking of external id. But I cant see how this can happen.

**Comment by abdallamhafeez** [comment]
@kittiu You can create a fresh database then add some expenses after that try to install a module

**Comment by kittiu** [comment]
OK, this make sense now. May be we didn't test with such case yet. Thank you.
CC @kongrattapong

**Comment by kittiu** [comment]
Fixed per #676 would you help review the PR?
cc @johnotoole68 @abdallamhafeez

**Comment by LeaderMalang** [comment]
I am also facing same problem with existing database.But when i had tried to install this module with newly created database then error doesn't occur.So if anyone have solution then please let me know.

**Comment by kittiu** [comment]
It is fixed in https://github.com/OCA/hr/pull/746
Is it not?

**Comment by LeaderMalang** [comment]
Fixed thanks

**Comment by Saran440** [comment]
Fixed and merged #746
